### PR TITLE
Feature/seperate pd amounts

### PR DESCRIPTION
**Pull request by Sandarr95**
## Issue

Please link to an existing issue (or open one before submitting a PR).

## Tasks

 - [ ] ...
 - [ ] Update `docs/API.md` (remove if irrelevant)
 - [ ] Update `CHANGELOG.md` (remove if irrelevant)
 - [ ] review

### CODE REVIEW

File: src/structures/vroom/input/input.cpp [hunk 1]
```diff
@@ -106,9 +106,6 @@ void Input::add_shipment(const Job& pickup, const Job& delivery) {
   if (pickup.priority != delivery.priority) {
     throw Exception(ERROR::INPUT, "Inconsistent shipment priority.");
   }
-  if (!(pickup.pickup == delivery.delivery)) {
-    throw Exception(ERROR::INPUT, "Inconsistent shipment amount.");
-  }
```
Comment (jcoupey) [inline src/structures/vroom/input/input.cpp:194]: Needs to be removed indeed to allow "asymmetric" amounts for pickup and deliveries.

File: src/structures/vroom/raw_route.cpp [hunk 2]
```diff
@@ -69,7 +69,7 @@ void RawRoute::update_amounts(const Input& input) {
       current_nb_pickups += 1;
       break;
     case JOB_TYPE::DELIVERY:
-      assert(job.delivery <= current_pd_load);
```
Comment (jcoupey) [inline src/structures/vroom/raw_route.cpp:72]: Makes sure we don't end up in a situation where we deliver more than what's been picked along the route. You should be able to keep this.

File: src/structures/vroom/raw_route.cpp [hunk 3]
```diff
@@ -80,7 +80,7 @@ void RawRoute::update_amounts(const Input& input) {
     _nb_pickups[i] = current_nb_pickups;
     _nb_deliveries[i] = current_nb_deliveries;
   }
-  assert(_pd_loads.back() == input.zero_amount());
```
Comment (jcoupey) [inline src/structures/vroom/raw_route.cpp:83]: Makes sure everything picked along the route has been delivered somehow. OK for removal.

File: src/utils/helpers.h [hunk 4]
```diff
@@ -590,7 +590,7 @@ inline Route format_route(const Input& input,
     current_load -= current_job.delivery;
     sum_pickups += current_job.pickup;
     sum_deliveries += current_job.delivery;
-    assert(current_load <= v.capacity);
```
Comment (jcoupey) [inline src/utils/helpers.h:719]: Should be enforced by the optimization search, you should be able to keep this.

**Review by jcoupey — COMMENTED** [review]
The Job ctor will populate the `Job::pickup` and `Job::delivery` member values differently depending on if a step is a`JOB_TYPE::PICKUP` or `JOB_TYPE::DELIVERY`. I think you need to adjust this, see [my previous comment](https://github.com/VROOM-Project/vroom/issues/329#issuecomment-624754573).

**Comment by Sandarr95** [comment]
@jcoupey I've been looking into this again. The comments about the asserts seem to be correct, thanks for that. The optimization search is not adhering to it, unfortunately. It will still propose routes exceeding capacity, as if it "expects" every delivery will decrease the capacity load. Keeping the "assert" to make sure the `current_load` is under the `v.capacity` then aborts the program with an error.

Reading into this more I saw over the past year [#545](https://github.com/VROOM-Project/vroom/pull/545) was implemented. It gave me good directions on where to start with modifying the optimization search.

I did want to ask though if you still see benefit in implementing this, as it would only expand the functionality gained with [#545](https://github.com/VROOM-Project/vroom/pull/545) by allowing multiple dimensions on it. Although, I can't say (yet) if it might also be better optimized.

**Comment by jcoupey** [comment]
Looking back at https://github.com/VROOM-Project/vroom/issues/329, it seems like your need was to limit the number of tasks per vehicle, exactly what https://github.com/VROOM-Project/vroom/issues/421 is about. Now that's it's available and released, I don't really see the benefit of the initial workaround any more?

**Pull request closed without merging**